### PR TITLE
Implement gas counting without system call

### LINKED ISSUE #1468: Make gas metering without a syscall
see https://github.com/paritytech/wasm-instrument/issues/11

- [x] #1604
- [x] #1605
- [x] #1606
- [x] #1607

### PR DESCRIPTION
**Pull request by gshep**
Resolves #1468 

Blocked by #1689 , https://github.com/gear-tech/substrate/pull/23

- refactor syscalls `create_program*` in wasmi backend
- fix regression in `exit` syscall for wasmi backend
- extend `Ext` trait: getters/setters for counters, `out_of_...` methods
- remove old `gas` handler
- use new injection method
- add `ext_sandbox_set_global_val_version_1` to the list of allowed imports in `wasm-proc`
- adjust other source code base to the changes

**Release notes**: Previous implementation of gas metering calls into the supervisor for every basic block in order to report some consumption from the gas meter. This is slow because calling imported functions is a heavy operation.
This PR contains implementation of gas metering in programs' userspace for both wasmer and wasmi execution backends. Userspace gas metering reduces gas consumption by programs significantly. Several tiny bugs were also fixed with these changes.

@gear-tech/dev

### CODE REVIEW

File: core-backend/wasmi/src/env.rs [hunk 1]
```diff
@@ -56,6 +55,10 @@ pub enum WasmiEnvironmentError {
     PreExecutionHandler(String),
     #[from]
     StackEnd(StackEndError),
+    #[display(fmt = "Gas counter not found or has wrong type")]
```
Comment (breathx) [inline core-backend/wasmi/src/env.rs:58]: Can't here appear error `Mutable globals are not supported`?

File: core-processor/src/ext.rs [hunk 2]
```diff
@@ -689,6 +685,43 @@ impl EnvExt for Ext {
...
+        )
+    }
+
+    fn update_counters(&mut self, gas: u64, allowance: u64) {
+        let gas_left = self.context.gas_counter.left();
+        if gas_left > gas {
+            self.context.gas_counter.charge(gas_left - gas);
+        } else {
+            self.context.gas_counter.refund(gas - gas_left);
+        }
```
Comment (breathx) [inline core-processor/src/ext.rs:702]: probably check for equality and do nothing?

File: pallets/gear/src/lib.rs [hunk 3]
```diff
@@ -629,7 +629,7 @@ pub mod pallet {
                 code,
                 schedule.instruction_weights.version,
                 Some(module),
-                false,
+                true,
```
Comment (breathx) [inline pallets/gear/src/lib.rs:632]: there is also 2 usage of `Code::new_raw` and shouldn't they also be with true flag?

**Review by breathx — APPROVED** [review]
Magnificent work!

File: pallets/gear/src/lib.rs (hunk 3 above)
Comment (gshep) [reply pallets/gear/src/lib.rs:632]: @ukint-vs please take a closer look to this

File: core-backend/wasmi/src/env.rs (hunk 1 above)
Comment (gshep) [reply core-backend/wasmi/src/env.rs:58]: Nope, since wasmi has been updated to 0.14.0

File: pallets/gear/src/lib.rs (hunk 3 above)
Comment (breathx) [reply pallets/gear/src/lib.rs:632]: Discussed with @ukint-vs, we will deal with all of related to `_raw` calls after benchmark reworking

Comment (ukint-vs) [reply pallets/gear/src/lib.rs:632]: we can merge it as it is and refactor benchmarks later with @grishasobol

Comment (gshep) [reply pallets/gear/src/lib.rs:632]: ok

**Review by grishasobol — APPROVED** [review]

File: core-processor/src/ext.rs (hunk 2 above)
Comment (gshep) [reply core-processor/src/ext.rs:702]: I would leave it as is

**Pull request merged**